### ISSUE TITLE
Finite state machine base

### ISSUE THREAD

**Issue by habanoz** [tezos-reward-distributor-organization/tezos-reward-distributor#392]
Using FSM based approach is proposed. TRD needs building blocks for implementing FSM-based operations.

**Comment by habanoz** [comment]
"fysom" is simple and seems to offer enough features. "transitions" is newer, more active, and offers more features. "fysom" is already in the requirements list because it was used in configure.py, thus "fysom" is more preferable. 

FsmBuilder is added for more expressive FSM building.
TrdFsmModel is added as a proxy around "fysom" fsm model object to decrease coupling. Also, it may help to switch FSM library without much effort, in the future if "fysom" fail to meet our needs.

**Comment by vkresch** [comment]
@habanoz I have a couple of questions concerning the PR which are unclear to me:
1) Which third party for FSM should be used? `transitions` or `fysom`? Because currently both are used. 
2) Why are you implementing the transition methods and states yourself if you could use the builtin methods from the external libraries but just with custom callback functions for it like in the [quick start](https://github.com/pytransitions/transitions#quickstart)?

For example:
Instead of defining the add_transition method from fsm_builder:
```
fsm_builder.add_transition(TrdEvent.LAUNCH, TrdState.INITIAL, TrdState.CMD_ARGS_PARSED)
```

Using the builtin `add_transition` method from Machine with the states included (simplified example):
```
from transitions import Machine
machine = Machine(......)
machine.add_transition(TrdEvent.LAUNCH TrdEvent.INITIAL, TrdState.CMD_ARGS_PARSED)
...
```

There are also some builtin arguments to handle the entry and exit of a state and even conditions. 
Thanks in advance. Btw I am still familarizing myself with the source code of the project so sorry if I missed something.

**Comment by habanoz** [comment]
@vkresch 

There are some comments in 393 which are related to your questions. Let me provide a collective summary here.

> Which third party for FSM should be used? transitions or fysom? Because currently both are used.

I started with "fysom" because of the following reasons:
- It was used in configure.py and already among dependency list.
- It is simpler, at least seems to be. Provides enough features to be used. Transitions on the other hand comes with more features which means it is heavier. 

Then I switched to "Transitions" for the following reason:
- It turned out Fysom is not suitable for our needs. It comes with two usage options: regular fsm and global fsm. Regular fsm does not support conditions while global fsm is not allowing coexistence of two different state machines at the same time when used in a common model. Why we used a common model is discussed below.

This is why we have two frameworks implemented. Last night I removed the unused fysom based classes because now it is clear we will continue with "transitions" library.

> Why are you implementing the transition methods and states yourself if you could use the builtin methods from the external libraries but just with custom callback functions for it like in the quick start?

It is because we do not want tight coupling to third party libraries. I added two abstraction classes: 
- TrdFsmBuilder: Hides details about fsm building. 
- TrdFsmModel: Hides the fsm object. 

Our case is a perfect showcase to demonstrate why abstractions around third party libs should be used. I managed to change the underlying framework without changing the whole code. And there is no guaranty that we will not need to change the lib again. 

Second reason is the given libraries are general purpose libraries. They come with bunch of features which TRD does not need. Having many features makes learning hard and leads to errors. TrdFsmModel abstraction limits the features to a set which contains only the necessary features. 

When a new developer tries to add a new state he will not need to learn about transitions library. He will be able to do this by using a small set of functions in TrdFsmModel abstraction which are self explanatory.
